### ISSUE TITLE
Option to only export deform bones

### ISSUE THREAD

**Issue by akien-mga** [godotengine/collada-exporter#14]
Moved from https://github.com/godotengine/godot/issues/4074:

@Lamoot wrote:

> Testing the Better Collada Exporter for Blender, I notice all of the bones in a rig get exported to Godot. A desired way would be to only export those bones that directly influence the mesh (i.e. deform bones). In Blender, bones can be flagged as being 'deform' bones and e.g. the FBX exporter has an option to only export those. It would be nice and very much needed for Better Collada to support this.
> 
> A rig is usually made from many bones. There are deform bones which directly deform the mesh, then there are control bones that are used to animate, and also helper bones. Together they make a rig comfortable to use and animate. However, for games, you only need the deform bones, the rest can and should be discarded.

**Comment by avril-gh** [comment]
blenders rigify could be a good example.

**Comment by Lamoot** [comment]
As an example of a rig setup, I'm attaching a simple stretchy cube. The rig is made up of control, helper, and deform bones. The archive contains the .blend and .dae files exported with the better collada exporter.
![stretchy_cube](https://cloud.githubusercontent.com/assets/3763179/17085086/bde22676-51cf-11e6-9175-2fa9d7b255ab.png)
[stretchy_cube.zip](https://github.com/godotengine/collada-exporter/files/380374/stretchy_cube.zip)

**Comment by RodeoMcCoy** [comment]
For what it's worth, I would really love to see this happen still. I've looked at the code, but it's a little dense for me now, I need to brush up on my Python.

Rigify is an essential part of my toolkit, and unless I'm mistaken it's becoming more popular with it being in included in Blender now.

**Comment by avril-gh** [comment]
> Rigify is an essential part of my toolkit

same here

**Comment by set-killer** [comment]
You know that there is this option:
```
- Exclude Control Bones -
Exclude skeleton bones with names beginning with 'ctrl'.
```
Is this going to give you the same result as exporting only the deform bones?

If this does not works then a similar functionality should be implemented, based on this Blender' property:
https://docs.blender.org/manual/en/dev/rigging/armatures/bones/properties/deform.html
So if we go with the `Deform` checkbox from the blender's properties you should check and uncheck the bones which you need before exporting, and back again check/ucheck the ones you need when animating, rigging, whatever...

**Comment by RodeoMcCoy** [comment]
I've actually been playing with those properties today, to no avail.

Looking at the code, it appears the "Exclude control bones" option does a simple string comparison of the bone's object name (it checks if the bone's name is prefixed with "ctrl-" without the quotes). This would be sufficient for Rigify and also the Makehuman MHX rig and probably others as the convention is to name deforming bones with the prefix "DEF-". 

Checking the blender property would be the better way to do it, as then it is portable to custom rigs which may not follow the naming convention above. That requires a knowledge of the Blender API which is above me right now.

**Comment by set-killer** [comment]
@RodeoMcCoy  How do you specify which bone is deform bone and which one is control bone? Is it by that property from the link above?

Because:
> Also it is worth noting that by turning off a bone’s deform option, makes it not influence the mesh at all...

I guess both the control bones and the deform bones should modify the mesh, so using the property from the link above is not a good option i suppose?

From the attached example by @Lamoot  none of the bones have this property checked.

------------
By using the prefix you can specify exactly which bone you want to be exported...

**Comment by Lamoot** [comment]
Greetings!

In Blender to set a bone as deform, you can do it in the **Properties** window, **Bone** category, **Deform** panel. Of course, you need to have the bone selected. You might not have selected the deform bone if it doesn't show deform option enabled, as there are other bones in that space.

![set_deform](https://cloud.githubusercontent.com/assets/3763179/22695175/1f65c54a-ed4a-11e6-8d4f-744789310dac.png)

I'm aware of the current option to discard bones with ctrl prefix, but it's limiting to do it this way. For example in rig hierarchy, some bones are neither for deform nor control, but sit in between to automate parts of the rig. It's why I'm proposing to base this on a bone's property and let the rigger enable / disable deform for desired bones, independently of any naming convention. Below you can see python output when toggling the option on / off.

![set_deform1](https://cloud.githubusercontent.com/assets/3763179/22695749/1abce53a-ed4c-11e6-8645-5cd460aa6d7c.png)

**Comment by RodeoMcCoy** [comment]
Lamoot thanks for the great answer. set-killer, thanks for your commit! I took a look at your code changes and it's simple enough and should do the trick. I can submit an approval if you'd like. When it's merged I'll be excited to test it out.

**Comment by set-killer** [comment]
You may test it by downloading it as a [raw file](https://raw.githubusercontent.com/set-killer/collada-exporter/701f5a4bbc8ab8403897e3ccf6fdc0f856c44f90/io_scene_dae/export_dae.py) and replace the script which you are currently using.

**Comment by RodeoMcCoy** [comment]
Works like a charm! I only tested with one model rigged with Rigify. I previously had animation issues (wonky mesh deforms that were terrible) and the Skeleton node in Godot had ~430 bones. Now with these changes all the animations are perfect and the Skeleton node has 77 bones.

Thanks!

**Comment by Lamoot** [comment]
Yup! The option works here as well with the various files I tested. It does one thing a bit differently in that it doesn't discard the non-deform root bone, though I don't think this gets in the way. Thank you very much for implementing this.

Do you think the tooltip / name of the option should now be adjusted to better communicate what it does? Here's an example from the fbx script. 
![set_deform2](https://cloud.githubusercontent.com/assets/3763179/22831338/43702220-efaa-11e6-8b5f-8f81bdce38d0.png)

**Comment by set-killer** [comment]
The tooltip will be changed to:

> Exclude skeleton bones with names beginning with 'ctrl' or bones which are not marked as Deform bones.

Because we are still keeping the old functionality. The changes are in the merge request, but outside of the file which you replaced and that actually contains the logic.

So we are waiting for merging. Thank you for your cooperation!

**Comment by Lamoot** [comment]
@set-killer The tooltip sounds good. Happy to help you getting this much needed addition to the exporter.

**Comment by RodeoMcCoy** [comment]
@set-killer It's the default Rigify one. It's more for film animation rather than games, but it's such a nice tool. The drawback was the high number of bones, which game engines don't like.

**Comment by akien-mga** [comment]
Fixed by #45.
